### PR TITLE
Overrides module refactor

### LINKED ISSUE #2641: Clean up config in overrides module
The overrides module has outgrown its original scope of providing limits to Tempo components, to hosting any kind of dynamic configuration that can be modified at runtime. So, in the same config block we have ingestion limits like `ingestion_rate_limit_bytes` and `max_traces_per_user`, metrics-generator config such as `metrics_generator_ring_size` and `metrics_generator_processor_service_graphs_histogram_buckets`, and soon even storage config with the introduction of [`vParquet3`](https://github.com/grafana/tempo/pull/2551).

All of this makes for a complex config structure, with overly long param names (eg. `metrics_generator_processor_service_graphs_enable_client_server_prefix`), and unclear scope.

### Proposal
- [ ] Add indentation to overrides block to make reading and writing configs easier

    Old overrides:
    ```yaml
    overrides:
        ingestion_rate_strategy: global
        ingestion_rate_limit_bytes: 1000000
        ingestion_burst_size_bytes: 1000000
        max_traces_per_user: 10000
        metrics_generator_ring_size: 3
        metrics_generator_processors: [span-metrics, service-graphs]
        metrics_generator_max_active_series: 100000
        metrics_generator_processor_service_graphs_dimensions: [http.status, http.method]
        metrics_generator_processor_span_metrics_enable_target_info: true
    ```

    New overrides:
    ```yaml
    overrides:
        ingestion:
            rate_strategy: global
            rate_limit_bytes: 1000000
            burst_size_bytes: 1000000
            max_traces_per_user: 10000
        metrics_generator:
            ring_size: 3
            processors: [span-metrics, service-graphs]
            max_active_series: 100000
            processor:
                service_graphs:
                    dimensions: [http.status, http.method]
                span_metrics:
                    enable_target_info: true
    ```

### PR DESCRIPTION
**Pull request by mapno**
**What this PR does**:

Follow up to https://github.com/grafana/tempo/pull/2638

The [`Limits` struct](https://github.com/grafana/tempo/blob/a5a7adb5cdd593650650aa870d316c6b4599b354/modules/overrides/limits.go#L47-L106) has a mixed responsibility:

* it contains all the limits (we have more than 30 now!)
* it contains configuration for the overrides module itself

When it is used in the top-level `App` struct it contains default values for the limits and configuration for per-tenant overrides. But when `Limits` is used by the runtime config manager it contains per tenant values of the limits.

The dual purpose of `Limits` is awkward and confusing.

Furthermore, the overrides module has outgrown its original scope of providing limits to Tempo components, to hosting any kind of dynamic configuration that can be modified at runtime. In the same config block we have ingestion limits like `ingestion_rate_limit_bytes`, metrics-generator config such as `metrics_generator_ring_size`, and (soon™️) storage config with the introduction of https://github.com/grafana/tempo/pull/2551.

All of this makes for a complex config structure, with overly long param names (eg. `metrics_generator_processor_service_graphs_enable_client_server_prefix`), and unclear scope.

This PR addresses this by creating a `Config` struct that is explicitly intended to configure the overrides module, ensuring the `Limits` struct can focus on holding limits only. It also adds indentation to the parameters, making them easier to read and write.

These has effects on the tempo config file and the per tenant overrides file.

Old config:
```yaml
overrides:
  ingestion_rate_strategy: local
  ingestion_rate_limit_bytes: 12345
  ingestion_burst_size_bytes: 67890
  max_search_duration: 17s
  forwarders: ['foo']
  metrics_generator_processors: [service-graphs, span-metrics]
```

New config:
```yaml
overrides:
  defaults:
    ingestion:
      rate_strategy: local
      rate_limit_bytes: 12345
      burst_size_bytes: 67890
    read:
      max_search_duration: 17s
    forwarders: ['foo']
    metrics_generator:
      processors: [service-graphs, span-metrics]
```

**Which issue(s) this PR fixes**:
Fixes #2641

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: modules/overrides/config.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+)
+
+var metricLimitsDesc = prometheus.NewDesc(
+	"tempo_limits_defaults",
+	"Default resource limits",
+	[]string{"limit_name"},
+	nil,
+)
+
+type IngestionConfig struct {
```
Comment (kvrhdn) [inline modules/overrides/config.go:54]: What do you think about calling these structs `IngestionOverrides` instead? The only true configuration struct is `Config`, all the other structs contain overrides.

File: modules/overrides/config.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	// Distributor enforced overrides.
+	RateStrategy   string `yaml:"rate_strategy,omitempty" json:"rate_strategy,omitempty"`
+	RateLimitBytes int    `yaml:"rate_limit_bytes,omitempty" json:"rate_limit_bytes,omitempty"`
+	BurstSizeBytes int    `yaml:"burst_size_bytes,omitempty" json:"burst_size_bytes,omitempty"`
+
+	// Ingester enforced overrides.
```
Comment (kvrhdn) [inline modules/overrides/config.go:60]: I'd keep the original comments: these are still actual limits we enforce, these limits are just part of the overrides (= properties we can configure at run-time).

```suggestion
	// Distributor enforced limits.
	RateStrategy   string `yaml:"rate_strategy,omitempty" json:"rate_strategy,omitempty"`
	RateLimitBytes int    `yaml:"rate_limit_bytes,omitempty" json:"rate_limit_bytes,omitempty"`
	BurstSizeBytes int    `yaml:"burst_size_bytes,omitempty" json:"burst_size_bytes,omitempty"`

	// Ingester enforced limits.
```

**Review by kvrhdn — COMMENTED** [review]
Nice work! This must have been a ton of work to track down all the compilation errors 😅

**Review by knylander-grafana — COMMENTED** [review]
Docs look good. Thank you for adding them.

File: cmd/tempo/app/config.go [hunk 3]
```diff
@@ -49,7 +49,7 @@ type Config struct {
 	Ingester        ingester.Config         `yaml:"ingester,omitempty"`
 	Generator       generator.Config        `yaml:"metrics_generator,omitempty"`
 	StorageConfig   storage.Config          `yaml:"storage,omitempty"`
-	LimitsConfig    overrides.Limits        `yaml:"overrides,omitempty"`
+	OverridesConfig overrides.Config        `yaml:"overrides,omitempty"`
```
[the remark this answers is not in the document]
Comment (mapno) [reply cmd/tempo/app/config.go]: Changed in 0de405260025ee7b3eb7ab66339f73a05d0a459c

File: modules/overrides/config.go (hunk 1 above)
Comment (mapno) [reply modules/overrides/config.go]: Agree. Done in 0de405260025ee7b3eb7ab66339f73a05d0a459c

File: modules/overrides/config.go (hunk 2 above)
Comment (mapno) [reply modules/overrides/config.go]: Right. Done in 0de405260025ee7b3eb7ab66339f73a05d0a459c

File: modules/overrides/config.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	// MetricsGenerator enforced overrides.
+	MetricsGenerator MetricsGeneratorConfig `yaml:"metrics_generator,omitempty" json:"metrics_generator,omitempty"`
+	// Forwarders
+	Forwarders []string `yaml:"forwarders,omitempty" json:"forwarders,omitempty"`
+	// Global enforced overrides.
+	Global GlobalLimitsConfig `yaml:"global,omitempty" json:"global,omitempty"`
+}
+
+type Config struct {
+	DefaultOverrides Overrides `yaml:"defaults,omitempty" json:"default,omitempty"`
```
[the remark this answers is not in the document]
Comment (mapno) [reply modules/overrides/config.go]: Ups, nice catch 0de405260025ee7b3eb7ab66339f73a05d0a459c

File: modules/overrides/config.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+	// Configuration for overrides module, convenient if it goes here.
```
[the remark this answers is not in the document]
Comment (mapno) [reply modules/overrides/config.go]: :yoohoo: 0de405260025ee7b3eb7ab66339f73a05d0a459c

File: modules/overrides/config.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+		return nil
+	}
+
+	// Try to unmarshal inline limits
+	type legacyConfig struct {
+		DefaultOverrides LegacyOverrides `yaml:",inline"`
+
+		PerTenantOverrideConfig string         `yaml:"per_tenant_override_config"`
+		PerTenantOverridePeriod model.Duration `yaml:"per_tenant_override_period"`
+	}
```
[the remark this answers is not in the document]
Comment (mapno) [reply modules/overrides/config.go]: Uh, nice one. Fixed in 0de405260025ee7b3eb7ab66339f73a05d0a459c

File: modules/overrides/legacy.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
+package overrides
```
[the remark this answers is not in the document]
Comment (mapno) [reply modules/overrides/legacy.go]: Good suggestion. Done in 0de405260025ee7b3eb7ab66339f73a05d0a459c

**Review by kvrhdn — COMMENTED** [review]
Looks good to me ✅

Can't approve since I also have changes in here though + I think having another review to make sure the new overrides struct is reasonable would be nice.

**Review by zalegrala — COMMENTED** [review]
This looks pretty good to me.  A little tough to review the `toLegacy` bits due to the high number of config options.  Have we been able to test this out somewhere?  A little nervous to merge on a Friday.

**Comment by mapno** [comment]
> Have we been able to test this out somewhere?

I've tested it locally, but I can deploy it somewhere if necessary.

**Review by zalegrala — APPROVED** [review]
This looks reasonable to me.  Do we have thoughts about how to transition our jsonnet to the new approach?

**Review by stoewer — APPROVED** [review]
The code changes look good to me
But I agree with @kvrhdn that the [docs for the overrides should be updated](https://github.com/grafana/tempo/pull/2688#pullrequestreview-1589416512). I think it would be OK to do this in an extra PR though

**Comment by mapno** [comment]
Hi folks, I've updated all the docs. I'll open a new issue to handle the migration of jsonnet.

**Comment by YayBurritos** [comment]
@mapno : How/where do I put this in the Tempo Distributed Helm chart? Currently working with the `1.7.1` chart (multitenancy not enabled). Thanks!

```
overrides:
  defaults:
    ingestion:
      rate_strategy: local
      rate_limit_bytes: 12345
      burst_size_bytes: 67890
    read:
      max_search_duration: 17s
    forwarders: ['foo']
    metrics_generator:
      processors: [service-graphs, span-metrics]
```

**Comment by mapno** [comment]
Hey. Unfortunately, this PR is required to use the new format https://github.com/grafana/helm-charts/pull/2825. Let's see if we can get a final approval.

**Comment by YayBurritos** [comment]
> Hey. Unfortunately, this PR is required to use the new format [grafana/helm-charts#2825](https://github.com/grafana/helm-charts/pull/2825). Let's see if we can get a final approval.

Thanks for the update! Didn't realize the Tempo Distributed Helm chart wasn't equipped to handle the new overrides format (aside from me trying to force it to happen). I may stick with the 'legacy' format for now in order to keep things moving on our side.

**Comment by ToonTijtgat2** [comment]
Dear

Can someone please give an example on how to use this new overrides config?

I placed it in the conf/tempo.yaml
Placed it on runtime-config/overrides.yaml
placed it under conf/overrides.yaml

It always keeps saying defaults is not found on the legacyoverrides.

The config I try to accomplice looks like this.
```overrides.yaml: |
    defaults:
      global:
        max_bytes_per_trace: 1500000
      metrics_generator:
        processors: [service-graphs, span-metrics]
        remote_write_headers:
          tenant-id: prometheus-apps-app-dev```

I have a feeling I oversee something, because it can not be that hard to get those things done I guess?

**Pull request merged**